Add web terminal handlers for file transfer requests

avatus opened the pull request:
Part of https://github.com/gravitational/teleport/pull/23546
Relies on https://github.com/gravitational/teleport/pull/24580

web buddy: https://github.com/gravitational/teleport/pull/24583

This PR adds the `fileTransferC` channel to our web terminal to listen to websocket events dealing with file transfer requests. The two events added are for requesting a new file transfer, and responding to an existing file transfer.

avatus committed:
Add file transfer handlers to web terminal

avatus committed:
remove unused keys

ravicious commented on lib/web/terminal.go:
```diff
@@ -394,7 +394,8 @@ func (t *TerminalHandler) handler(ws *websocket.Conn, r *http.Request) {
 	// communicate over the websocket.
 	resizeC := make(chan *session.TerminalParams, 1)
-	stream, err := NewTerminalStream(ws, WithTerminalStreamResizeHandler(resizeC))
+	fileTransferC := make(chan *session.FileTransferParams)
```
Yeah, I assume this is heavily based on the existing code for `resizeC` right? If so I suppose it'd only make sense to use a buffered channel as well, no?

avatus replied:
mistype, ty for the catch!

ravicious commented on api/observability/tracing/ssh/session.go:
```diff
@@ -342,3 +342,47 @@ func (s *Session) CombinedOutput(ctx context.Context, cmd string) ([]byte, error
+func (s *Session) FileTransferRequestResponse(ctx context.Context, req FileTransferResponseReq) error {
+	const request = "file-transfer-request-response"
+	config := tracing.NewConfig(s.wrapper.opts)
+	ctx, span := config.TracerProvider.Tracer(instrumentationName).Start(
```
Do we need to set up all of this tracing stuff even though in the end we end up just calling `s.Session.SendRequest` which does that again? I don't really know much about how we do tracing, I'm looking at `SetEnv` which ends up calling `SendRequest` too, but `SetEnv` actually does some extra stuff.

avatus replied:
`s.Session.SendRequest` doesn't have the tracing as that uses the wrapped `ssh.Session`, but if I instead just used `s.SendRequest`, then yeah, I can ignore the tracing as _that_ `SendRequest` does set it up. Good call

ravicious replied:
Ah, I didn't notice the difference between the two. Still, I'd recommend asking on #dev-teleport if one way of doing this is preferred over the other.

avatus commented on lib/web/terminal.go:
```diff
@@ -853,6 +857,33 @@ func (t *TerminalHandler) streamEvents(tc *client.TeleportClient) {
+			return
+		case transferRequest := <-fileTransferC:
+			// if not response, this is a new file transfer request
+			if !transferRequest.Response {
```
Yeah, this grew from "eh only 1 field is needed" to a few, and I didn't cut them apart. It's best to do that now and make it more clear.

avatus replied:
After splitting the structs, I made a new `chan`, one to handle the requests, and one to handle the responses. The original reason to have this struct be the same was to reuse the same `chan` and just make decisions based on what info came through but I think thats a bit lazy now. Although I have a separate channel, I kept the "setting" of the channels in the same method (just with the two channels now instead of one) and also, have both cases in the same handler as well. Came out cleaner and easier to parse around throughout the app. Thanks for keeping me accountable

avatus committed:
Make fileTransferC buffered

avatus commented on lib/srv/termhandlers.go:
```diff
@@ -145,11 +145,10 @@ func (t *TermHandlers) HandleFileTransferRequestResponse(ctx context.Context, ch
 		if err != nil {
 			return trace.Wrap(err)
 		}
-		t.SessionRegistry.NotifyFileTransferRequest(req, FileTransferApproved, scx)
```
This is just some base branch shenanigans. Both of the `Notify`s happen inside the approve/deny methods now but it'll be updated here when I get the base branch updated/merged. Ignore for now!

avatus committed:
Review feedback

avatus committed:
Fix typo

avatus commented:
Didn't mean to rebase this early and didn't realize after I pushed, sorry!

avatus commented on lib/web/terminal.go:
```diff
@@ -1168,6 +1208,40 @@ func (t *TerminalStream) Read(out []byte) (n int, err error) {
+		if err != nil {
+			return 0, trace.Wrap(err)
+		}
+		approved, ok := e["approved"].(bool)
```
This `approved` is handling the file transfer response. Is the response "Approved" or "Denied" basically.

avatus commented on api/observability/tracing/ssh/ssh.go:
```diff
@@ -58,6 +58,18 @@ type EnvsReq struct {
+	Direction string
+	Location  string
+	Filename  string
+	Size      string
```
Size shouldn't be here, I've removed it.

avatus committed:
Merge branch 'michaelmyers/ssh_server_file_transfer_handlers' into michaelmyers/web_terminal_file_request_handlers

avatus committed:
Merge branch 'michaelmyers/ssh_server_file_transfer_handlers' into michaelmyers/web_terminal_file_request_handlers

avatus committed:
Remove comment

avatus committed:
Merge branch 'michaelmyers/ssh_server_file_transfer_handlers' into michaelmyers/web_terminal_file_request_handlers

avatus committed:
Simplify session param

avatus committed:
Merge branch 'michaelmyers/ssh_server_file_transfer_handlers' into michaelmyers/web_terminal_file_request_handlers

avatus committed:
Simplify file transfer param callsite

ravicious commented on lib/web/terminal.go:
```diff
@@ -853,6 +857,33 @@ func (t *TerminalHandler) streamEvents(tc *client.TeleportClient) {
+		select {
+		case <-t.terminalContext.Done():
+			return
+		case transferRequest := <-fileTransferC:
```
Shouldn't the new version of `handleFileTransfer` account for individual channels getting closed?

ravicious commented on lib/web/terminal.go:
```diff
@@ -1168,6 +1208,49 @@ func (t *TerminalStream) Read(out []byte) (n int, err error) {
+		if err != nil {
+			return 0, trace.Wrap(err)
+		}
+		download, ok := e["download"].(bool)
+		if !ok {
+			return 0, trace.BadParameter("Unable to find approved status on response")
```
The error message doesn't match the check that's being performed here.

ravicious commented on lib/defaults/defaults.go:
```diff
@@ -662,6 +662,13 @@ const (
+
+	// WebsocketFileTransferRequestResponse is received when a response (approve/deny) has been
+	// made for an existing file transfer request
+	WebsocketFileTransferRequestResponse = "t"
```
It'd be nice to add some tests for that based on `TestResizeTerminal`.

avatus replied:
Added 👍

ravicious reviewed (approved):
Some minor adjustments are needed but looks ok otherwise.

ravicious commented on api/observability/tracing/ssh/session.go:
```diff
@@ -342,3 +342,21 @@ func (s *Session) CombinedOutput(ctx context.Context, cmd string) ([]byte, error
+
+// FileTransferRequestApprove sends a "file-transfer-request-response" ssh request
+// The response will contain an Approve bool which will approve or deny a requested file transfer
+func (s *Session) FileTransferRequestResponse(ctx context.Context, req FileTransferResponseReq) error {
```
When I was reviewing the web PR, I realized that this might need a better name. The other method names are verb based while this one feels a bit odd compared to them.

avatus replied:
I tried thinking of something that would make sense for both like `ResponseToFileTransferRequest` but, idk I didn't like it. So I just split them into to, `ApproveFileTransferRequest` and `DenyFileTransferRequest`. They both take just the `requestID` now as a string, and the req is created inside the method and sets `Approved` appropriately. Maybe a bit verbose but much more readable and understandable

avatus committed:
Merge branch 'michaelmyers/ssh_server_file_transfer_handlers' into michaelmyers/web_terminal_file_request_handlers

avatus committed:
Use api constants

avatus commented on api/observability/tracing/ssh/session.go:
```diff
@@ -342,3 +342,21 @@ func (s *Session) CombinedOutput(ctx context.Context, cmd string) ([]byte, error
+// RequestFileTransfer sends a "file-transfer-request" ssh request that will create a new file transfer request
+// and notify the parties in an ssh session
+func (s *Session) RequestFileTransfer(ctx context.Context, req FileTransferRequestReq) error {
+	const request = "file-transfer-request"
```
Done

avatus committed:
Split FileTransferRequestResponse methods

avatus committed:
Merge base

avatus committed:
Change name to FileTransferDecision

avatus committed:
Merge base

avatus committed:
Update ssh request name

avatus committed:
Merge branch 'michaelmyers/ssh_server_file_transfer_handlers' into michaelmyers/web_terminal_file_request_handlers

avatus committed:
use FileTransferReq

avatus committed:
Rename struct

avatus committed:
Add file transfer test

avatus committed:
Add FileTransferDecision to test

avatus committed:
Add dstPath to isApprovedFileTransfer

avatus committed:
Remove debug comment

avatus committed:
Merge branch 'michaelmyers/ssh_server_file_transfer_handlers' into michaelmyers/web_terminal_file_request_handlers

avatus committed:
Rename query params

avatus committed:
[web] Moderated Session file transfers (#24583)

* Add file transfer handlers to web terminal

* remove unused keys

* Add FileTransferContainer

This PR separates the absolute positioned container from the FileTransfer component. We will be adding FileTransferRequests
in the same spot, and this will allow them to be visible together without css shenanigans trying to absolutely place them together.
FileTransferRequests are separate from FileTransfers in that they interact with the tty.

* Add useFileTransfer hook

* Use useFileTransfer hook inside useSshSession

* Add tty handlers

* Update getScpUrl to add optional moderated params

* Use FileTransferContainer in Connect

* Update DocumentSSH

* remove fmt

* remove unused shellcmd key

* lint fixes

* remove pending file after get

* remove pending file after get

* Clean up console log

* Update filetransfererquest to use download bool

* Fix upload/download flow

* Fix typecheck and FileTransfer tests

* Fix UI display bugs and add license

* Change filetransferresponse to filetransferdecision

* Refactor useFileTransfer up to DocumentSsh

* Add license

* Fix lint error

* update pending transfer text

* use handleUpload/handleDownload, move FileTransferRequests to prop

* Review feedback

* Add usecallback to scpurls and useattempt

* Apply patch

* Update optional params

avatus committed:
Prefix file transfer env vars with TELEPORT_
